lib/zip/entry.rb

```diff
@@ -406,24 +406,28 @@ def get_extra_attributes_from_path(path) # :nodoc:
       @unix_uid   = stat.uid
       @unix_gid   = stat.gid
       @unix_perms = stat.mode & 0o7777
+      @time = ::Zip::DOSTime.from_time(stat.mtime)
     end
 
-    def set_unix_permissions_on_path(dest_path)
-      # BUG: does not update timestamps into account
+    def set_unix_attributes_on_path(dest_path)
       # ignore setuid/setgid bits by default.  honor if @restore_ownership
       unix_perms_mask = 0o1777
       unix_perms_mask = 0o7777 if @restore_ownership
       ::FileUtils.chmod(@unix_perms & unix_perms_mask, dest_path) if @restore_permissions && @unix_perms
       ::FileUtils.chown(@unix_uid, @unix_gid, dest_path) if @restore_ownership && @unix_uid && @unix_gid && ::Process.egid == 0
-      # File::utimes()
+
+      # Restore the timestamp on a file. This will either have come from the
+      # original source file that was copied into the archive, or from the
+      # creation date of the archive if there was no original source file.
+      ::FileUtils.touch(dest_path, mtime: time) if @restore_times
```

Comment:
Possibly one for a future PR... but we could use `File::utime` here as the original comment hinted, if we used the `atime` and `ctime` from the [`UniversalTime`](https://github.com/rubyzip/rubyzip/blob/8887b703490337f7ab68ebfb4880354db28ba648/lib/zip/extra_field/universal_time.rb#L10) extra field. (And we could store those from `File::stat`, too.)

Reply:
I think I used `mtime` because that is the only one `FileUtils.touch` supports. But now I look at the code for `FileUtils.touch` it turns out it uses `File.utime` internally anyway [1], setting `atime` and `mtime` to the same thing.

So, I agree, shall we leave this as is for now and look at it in a new PR?

[1] https://ruby-doc.org/stdlib-2.4.7/libdoc/fileutils/rdoc/FileUtils.html#method-c-touch

Reply:
Good points. I am definitely OK with that 👍.